ledger/metrics/src/main/scala/com/daml/metrics/grpc/GrpcMetricsServerInterceptor.scala

```diff
@@ -0,0 +1,150 @@
+      metrics.callsStarted.mark()
+      val metricsServerCall = new MetricsCall(
+        serverCall,
+        timerHandle,
+        metrics.messagesSent,
+        metrics.messagesSentSize,
+        metrics.callsFinished,
+        metrics.messagesRequested,
+      )
+      new MonitoringServerCallListener(
+        serverCallHandler.startCall(metricsServerCall, metadata),
+        metrics.messagesReceived,
+        metrics.messagesReceivedSize,
+      )
+    }
+  }
+
+  private final class MonitoringServerCallListener[T](
+      val delegate: Listener[T],
+      messagesReceived: Meter,
+      messagesReceivedSize: Histogram,
+  )(implicit
+      metricsContext: MetricsContext
+  ) extends ForwardingServerCallListener[T] {
+
+    override def onMessage(message: T): Unit = {
+      super.onMessage(message)
+      updateHistogramWithSerializedSize(messagesReceivedSize, message)
+      messagesReceived.mark()
+    }
+  }
+
+  private final class MetricsCall[ReqT, RespT](
+      call: ServerCall[ReqT, RespT],
+      timer: TimerHandle,
+      messagesSent: Meter,
+      messagesSentSize: Histogram,
+      callsClosed: Meter,
+      messagesRequested: Meter,
+  )(implicit metricsContext: MetricsContext)
+      extends SimpleForwardingServerCall[ReqT, RespT](call) {
+
+    override def request(numMessages: Int): Unit = {
+      super.request(numMessages)
+      messagesRequested.mark(numMessages.toLong)
+    }
+
+    override def sendMessage(message: RespT): Unit = {
+      super.sendMessage(message)
+      updateHistogramWithSerializedSize(messagesSentSize, message)
+      messagesSent.mark()
+    }
+
+    override def close(status: Status, trailers: Metadata): Unit = {
+      super.close(status, trailers)
+      withExtraMetricLabels(MetricsGrpcResponseCode -> status.getCode.toString) {
+        implicit metricsContext =>
+          callsClosed.mark()
+          timer.stop()
+      }
+    }
+  }
+
+  private def updateHistogramWithSerializedSize[T](
+      histogram: Histogram,
+      message: T,
+  )(implicit metricsContext: MetricsContext): Unit = {
+    message match {
+      case generatedMessage: JavaGeneratedMessage =>
+        histogram.update(generatedMessage.getSerializedSize)
+      case generatedMessage: ScalapbGeneratedMessage =>
+        histogram.update(generatedMessage.serializedSize)
+      case _ =>
+    }
+  }
+}
+object GrpcMetricsServerInterceptor {
+
+  val MetricsGrpcServiceName = "grpc_service_name"
+  val MetricsGrpcMethodName = "grpc_method_name"
+  val MetricsGrpcClientType = "grpc_client_type"
+  val MetricsGrpcServerType = "grpc_server_type"
+  val MetricsRequestTypeUnary = "unary"
+  val MetricsRequestTypeStreaming = "streaming"
+  val MetricsGrpcResponseCode = "grpc_code"
+
+}
+
+trait GrpcServerMetrics {
+  val callTimer: Timer
+  val messagesSent: Meter
+  val messagesSentSize: Histogram
+  val messagesReceived: Meter
+  val messagesReceivedSize: Histogram
+  val callsStarted: Meter
+  val callsFinished: Meter
```

Comment:
The `callsStarted` and `callsFinished` don't really work for me.
A client does _call_ a method on a service, but term 'call' is not really attached to gRPC.
I'd prefer something that reference methods, like `methodsInvoked` `methodsCalled`, `methodsCompleted`.

Reply:
I say that call is representative for what this signifies, the fact that it calls a method is irrelevant. In fact, `call` is part of the naming itself as it's the `C` in `gRPC`

Reply:
The [go gRPC interceptor](https://github.com/grpc-ecosystem/go-grpc-prometheus#labels) entirely skips `call` from its metric names (e.g., `grpc_server_started_total`/`grpc_server_handled_total`, `grpc_server_msg_sent_total`/`grpc_server_msg_received_total`). However, as the gRPC interceptors API even has `call` in the interface names (e.g., `ForwardingServerCallListener`, `ServerCallHandler`), I think it is an acceptable naming choice.

Reply:
Changed to drop the call from the name